### PR TITLE
Add RNA-seq UMAP example notebook

### LINKED ISSUE #128: Create UMAP dimension reduction example with RNA-seq data
Related to #110 

Currently we have a microarray example of UMAP. We should be able to largely copy the microarray example but switch out the dataset and the accompanying wording. 

Ideally we will use the same dataset as we will use for PCA RNA-seq example (#129) and ideally the dataset will not have too skewed of a global structure. 

This can be done separately from the "getting started" additions to each analysis #116

### PR DESCRIPTION
**Pull request by cbethell**
## Purpose:
What issue(s) does your PR address?

This PR closes #128 (the issue on creating a UMAP example notebook with RNA-seq data).

## Strategy
What was your strategy for this new or edited analysis?

As noted in one of my [comments on that issue](https://github.com/AlexsLemonade/refinebio-examples/issues/128#issuecomment-677895541), I followed a plan in the order as follows:

1. Read in `DESEq2`, `ggplot2` and `uwot` (now `umap`) packages 
2. Import and set up data - use same large non-QN’ed dateset as in PCA example notebook in PR #176
3. DESeq2- Create a `DESeqDatatset` object, define minimum counts cutoff, use `vst()` for normalization
4. Use `assay()` to extract the normalized data and run `uwot::umap()` -- this was substituted by the `umap::umap()` function as I had complications retaining the accession codes that are needed for joining with the metadata when using `uwot::umap()`
5. Plot using `ggplot()` — adding annotation labels one by one starting with 0 labels (and save using `ggsave()`)

## Concerns/Questions for reviewers:
What things should reviewers look out for?

This PR is still a work in progress (context and comments still need to be updated and some analysis steps added) so there are no particular questions for reviewers as yet other than:

- Are there suggestions for using `uwot::umap()` and retaining the accession codes so that the metadata can be joined or are we fine with using `umap::umap()` in this notebook?

### Content checks
* [x] All `{{BLANKS}}` have been replaced with the correct content.
* [x] [Sources are cited](https://github.com/AlexsLemonade/refinebio-examples/blob/master/CONTRIBUTING.md#citing-sources-in-text)
* [x] Seed is set (if applicable)

### Formatting Checks
* [x] Removed any manual numbering of sections.
* [x] Removed any instances of chunk naming.
* [x] Spell checked any Rmd file or md file.
* [x] Comments and documentation are up to date.

### Add datasets to S3
* [x] Added [data and metadata files to S3](https://github.com/AlexsLemonade/refinebio-examples/blob/master/CONTRIBUTING.md#adding-datasets-to-the-S3-bucket).

### Docker/Snakemake rendering components
* [x] Added the `.html` [link to the navigation bar](https://github.com/AlexsLemonade/refinebio-examples/blob/master/CONTRIBUTING.md#add-new-analyses-to-the-navbar).
* [x] Any not yet added packages needed for this analysis have been added to the Dockerfile and it successfully builds.
* [x] In the Docker container, [snakemake was run for rendering](https://github.com/AlexsLemonade/refinebio-examples/blob/master/CONTRIBUTING.md#how-to-re-render-the-notebooks).

### CODE REVIEW

**Comment by cansavvy** [comment]
One more thing that's a relatively quick change, the notebook names we need to make sure we have updated according to @jashapiro 's adjustment to the naming conventions: https://github.com/AlexsLemonade/refinebio-examples/blob/master/CONTRIBUTING.md#setting-up-a-new-analysis-file 

These means that `dimension_reduction` -> `dimension-reduction`

**Comment by cbethell** [comment]
> I think cautionary words/guidance should go in two places:
> 
> * Before the first umap calculation we should tell them a bit about what UMAP is and encourage them to run `?uwot::umap` and look at the parameters. Have a link out to a place that discusses parameters in more detail.
> * After they've printed out their last umap plot, we should have a mini section that discusses big picture thoughts about umap interpretation where the main point of it is ["Don't be too attached to these exact values that umap gave you"](https://github.com/AlexsLemonade/training-modules/blob/3dbc6f3f53c680ec6aa2f513851c1cd4635cc31c/scRNA-seq/05-dimension_reduction_scRNA-seq.Rmd#L382)

No problem! However, did you notice the note in my [original comment](https://github.com/AlexsLemonade/refinebio-examples/pull/193#issue-471897709) re `uwot::umap` versus `umap::umap` @cansavvy?


```
`uwot::umap()` -- this was substituted by the `umap::umap()` function
as I had complications retaining the accession codes that are needed for
joining with the metadata when using `uwot::umap()`
```

Do you have any suggestions on whether or not we should continue using `umap::umap()` or suggestions on how we can retain the accession codes for joining with the metadata using `uwot::umap`?

**Comment by cansavvy** [comment]
> Do you have any suggestions on whether or not we should continue using `umap::umap()` or suggestions on how we can retain the accession codes for joining with the metadata using `uwot::umap`?

I did not notice this. Lemme take a look into it.

File: 03-rnaseq/dimension_reduction_rnaseq_02_umap.Rmd [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+
+***
+
+# UMAP Visualization - RNA-seq
+
+## Install libraries
+
+See our getting started with [package installation](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#what-you-need-to-install) for a list of the other software you will need, as well as more tips and resources. 
+
+In this analysis, we will be using the R package [`DESeq2`](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) [@DESeq2] for normalization, the R package [`umap`](https://www.rdocumentation.org/packages/umapr/versions/0.0.0.9001/topics/umap) [@UMAP] for the production of UMAP dimension reduction values and the R package , and the R package [`ggplot2`](http://r-statistics.co/Complete-Ggplot2-Tutorial-Part1-With-R-Code.html) [@ggplot2] for plotting the UMAP values.
```
[the remark this answers is not in the document]
Comment (cbethell) [reply 03-rnaseq/dimension_reduction_rnaseq_02_umap.Rmd]: Will update these citation tags once #194 is merged (anticipating merge conflicts),

File: 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+TODO: Add a note/link to section in documentation on why we use `vst()` vs `rlog()` and `normTransform()`.
+
+```{r}
+# Normalize the data in the `DESeqDataSet` object using the `vst()` function from the `DESEq2` R package
+dds_norm <- vst(dds)
+```
+
+## Perform UMAP
+
+Uniform Manifold Approximation and Projection (UMAP) is a dimension reduction technique proposed by McInnes, Healy, & Melville as it isdescribed in the [associated paper](https://arxiv.org/abs/1802.03426) [@umap-paper].
```
Comment (cansavvy) [inline 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd:296]:
```suggestion
Uniform Manifold Approximation and Projection (UMAP) is a dimension reduction technique proposed by @umap-paper ([See associated paper](https://arxiv.org/abs/1802.03426)).
```

File: 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+While PCA assumes that the variation we care about has a particular distribution (normal, broadly speaking), UMAP allows more complicated distributions that it learns from the data. 
+The advantage of this feature is that UMAP can do a better job separating clusters, especially when some of those clusters may be more similar to each other than others.  
+
+In this code chunk, we are going to extract the normalized counts data from the `DESeqDataSet` object and perform UMAP on the normalized data using `umap()` from the `umap` package.
+We are using the default parameters when we run the `umap::umap()` function.
+Take a look at the `Tuning UMAP` section at [this link](https://cran.r-project.org/web/packages/umap/vignettes/umap.html) for more on parameters that can be configured when executing `umap::umap()` [@umap-parameters].
```
Comment (cansavvy) [inline 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd:302]:
```suggestion
Here's some [guidance about choosing parameters](https://cran.r-project.org/web/packages/umap/vignettes/umap.html#tuning-umap) when executing `umap::umap()` [@umap-parameters].
```

Comment (cansavvy) [reply 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd:302]: This citation will also need to be changed to `AuthorDate`

File: 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+# Now let's tell R to perform UMAP on the normalized data
+umap_results <- umap::umap(normalized_counts)
+```
+
+## Prepare data frame for plotting
+
+Now that we have the results from UMAP, we need to extract the counts data from the `umap_results` object and merge the variables from the metadata that we will use for annotating our plot.
+
+```{r}
+# We need to tell R to create a data frame with the umap counts and annotation data in preparation for plotting with `ggplot2`
```
Comment (cansavvy) [inline 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd:319]:
```suggestion
# We need to tell R to create a data frame with the umap values and annotation data in preparation for plotting with `ggplot2`
```

I don't think `counts` is the right word here (but I could be wrong).

File: 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+umap_plot_df <- data.frame(umap_results$layout) %>%
+  tibble::rownames_to_column("refinebio_accession_code") %>% # Let's get the rownames into a column so we can merge the annotation data
+  dplyr::inner_join(metadata, by = "refinebio_accession_code") # Let's merge the annotation data using the `refinebio_accession_code` column
```
Comment (cansavvy) [inline 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd:322]: I think there is a way to join using rownames, but honestly I find this strategy you used here easier to follow/explain.

File: 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+```
+
+## Create UMAP plot
+
+Now we can use the `ggplot` function to plot our normalized UMAP scores.
```
Comment (cansavvy) [inline 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd:327]:
```suggestion
Now we can use the `ggplot()` function to plot our normalized UMAP scores.
```

File: 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd [hunk 7]
```diff
@@ -0,0 +1,414 @@
...
+
+```{r}
+# Plot using `ggplot()` function
+ggplot(umap_plot_df, aes(x = X1, y = X2)) +
+  geom_point() # This tells R that we want a scatterplot
+```
+
+Let's try adding a variable to our plot for annotation.
+
+In this code chunk, the variable `refinebio_treatment` is given to the `ggplot()` function as part of the goal of the experiment is to analyze the sample transcriptional responses to androgen deprivation therapy (ADT).
```
Comment (cansavvy) [inline 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd:337]:
```suggestion
In this code chunk, the variable `refinebio_treatment` is given to the `ggplot()` function so we can label by androgen deprivation therapy (ADT) status.
```

File: 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd [hunk 8]
```diff
@@ -0,0 +1,414 @@
...
+  dplyr::inner_join(metadata, by = "refinebio_accession_code") # Let's merge the annotation data using the `refinebio_accession_code` column
+```
+
+## Create UMAP plot
+
+Now we can use the `ggplot` function to plot our normalized UMAP scores.
+
+```{r}
+# Plot using `ggplot()` function
+ggplot(umap_plot_df, aes(x = X1, y = X2)) +
```
Comment (cansavvy) [inline 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd:331]: Can you stretch this code chunk out same as you did below? I think that makes it easier for users to compare the ggplot runs.

File: 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd [hunk 9]
```diff
@@ -0,0 +1,414 @@
...
+
+```{r}
+# Plot using `ggplot()` function
+final_annotated_umap_plot <- ggplot(
+  umap_plot_df,
+  aes(
+    x = X1,
+    y = X2,
+    color = refinebio_treatment,
+    # Here we are giving `refinebio_treatment` to the color argument
```
Comment (cansavvy) [inline 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd:364]: Here I have the same comment from before: https://github.com/AlexsLemonade/refinebio-examples/pull/176#discussion_r474652655

Comment (cansavvy) [reply 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd:364]: My guess is you had worked on this before we got to that point on the PCA notebook. 😃

File: 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd [hunk 10]
```diff
@@ -0,0 +1,414 @@
...
+    shape = refinebio_disease # Here we are giving `refinebio_disease` to the shape argument
+  )
+) +
+  geom_point() # This tells R that we want a scatterplot
+
+# Display the plot that we saved above
+final_annotated_umap_plot
+```
+
+The plot above is more informative than the first plot we created but is still not as informative as one may home as there are no clear clusters represented.
```
Comment (cansavvy) [inline 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd:374]:
```suggestion
The plot above is more informative than the first plot we created but is still not as informative as one may home as there are no clear clusters represented.
```
"as one may home"? 😸 Honestly you can probably just delete this sentence anyway, not sure it adds much info.

File: 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd [hunk 11]
```diff
@@ -0,0 +1,414 @@
...
+Although it does appear that majority of the pre-ADT and post-ADT appear to cluster together, there are still questions remaining as we look at outliers.
+
+### Interpretation of UMAP plot and results
+
+1. Note that the coordinates of UMAP output for any given cell can change dramatically depending on parameters, and even run to run with the same parameters.
```
Comment (cansavvy) [inline 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd:379]:
```suggestion
1. Note that the coordinates of UMAP output for any given cell can change dramatically depending on parameters, and even run to run with the same parameters (Also why setting the seed is important). 
```

File: 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd [hunk 12]
```diff
@@ -0,0 +1,414 @@
...
+
+The plot above is more informative than the first plot we created but is still not as informative as one may home as there are no clear clusters represented.
+Although it does appear that majority of the pre-ADT and post-ADT appear to cluster together, there are still questions remaining as we look at outliers.
+
+### Interpretation of UMAP plot and results
+
+1. Note that the coordinates of UMAP output for any given cell can change dramatically depending on parameters, and even run to run with the same parameters. 
+This means that you should not rely too heavily on the exact values of UMAP's output. 
+
+  - One particular limitation of UMAP is that while observed clusters have some meaning, the distance *between* clusters usually does not (nor does cluster density). The fact that two clusters are near each other should NOT be interpreted to mean that they are more related to each other than to more distant clusters. (There is some disagreement about whether UMAP distances have more meaning, but it is also probably safer to assume they don't.)
```
Comment (cansavvy) [inline 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd:382]: I do have a question about whether we need to cite ourselves from training-modules? Not sure how this works.

File: 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd [hunk 13]
```diff
@@ -0,0 +1,414 @@
...
+# Save plot using `ggsave()` function
+ggsave(file.path(plots_dir, "SRP133573_umap_plot.png"), # Replace with name relevant your plotted data
+  plot = final_annotated_umap_plot
+)
+```
+
+# Resources for further learning
+
+- [More on UMAP and its parameters](https://cran.r-project.org/web/packages/umap/vignettes/umap.html) [@umap-parameters]
+- [Guidelines on choosing dimension reduction methods](https://journals.plos.org/ploscompbiol/article/file?id=10.1371/journal.pcbi.1006907&type=printable) [@dimension-reduction-guidelines]
```
Comment (cansavvy) [inline 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd:403]: These citations need to be updated too.

File: references.bib [hunk 14]
```diff
@@ -151,3 +151,25 @@ @Article{tidyverse
   pages = {1686},
   doi = {10.21105/joss.01686},
 }
+
+@Manual{UMAP,
```
Comment (cansavvy) [inline references.bib:155]:
```suggestion
@Manual{Konopka2020, 
```

File: references.bib [hunk 15]
```diff
@@ -151,3 +151,25 @@ @Article{tidyverse
...
+
+@Manual{UMAP, 
+  title = {Uniform Manifold Approximation and Projection}, 
+  author = {Tomasz Konopka}, 
+  note = {R CRAN version 0.2.6.0},
+  year = {2020}, 
+  url = {https://cran.r-project.org/web/packages/umap/umap.pdf}, 
+}
+
+@Article{umap-paper,
```
Comment (cansavvy) [inline references.bib:163]:
```suggestion
@Article{McInnes2018,
```

File: references.bib [hunk 16]
```diff
@@ -151,3 +151,25 @@ @Article{tidyverse
...
+  title = {UMAP: Uniform Manifold Approximation and Projection for Dimension Reduction},
+  author = {Leland McInnes and John Healy and James Melville},
+  year = {2018},
+  url = {https://arxiv.org/abs/1802.03426},
+}
+
+@Manual{umap-parameters,
```
Comment (cansavvy) [inline references.bib:170]:
```suggestion
@Manual{umap-vignette, 
```

File: 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd [hunk 17]
```diff
@@ -0,0 +1,414 @@
...
+
+***
+
+# UMAP Visualization - RNA-seq
+
+## Install libraries
+
+See our Getting Started page with [instructions for package installation](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#what-you-need-to-install) for a list of the other software you will need, as well as more tips and resources.
+
+In this analysis, we will be using the R package [`DESeq2`](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) [@DESeq2] for normalization, the R package [`umap`](https://www.rdocumentation.org/packages/umapr/versions/0.0.0.9001/topics/umap) [@UMAP] for the production of UMAP dimension reduction values and the R package , and the R package [`ggplot2`](http://r-statistics.co/Complete-Ggplot2-Tutorial-Part1-With-R-Code.html) [@ggplot2] for plotting the UMAP values.
```
Comment (cansavvy) [inline 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd:159]: These citations will also need to be updated with #194 changes.

File: 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd [hunk 18]
```diff
@@ -0,0 +1,414 @@
...
+# Attach the `ggplot2` library for plotting
+library(ggplot2)
+
+# We will need this so we can use the pipe: %>%
+library(magrittr)
+
+# Set the seed so our results are reproducible:
+set.seed(12345)
+```
+
```
Comment (cansavvy) [inline 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd:191]: Eventually I want to have things like `set.seed()` in some kind of glossary #117

File: 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd (hunk 12 above)
Comment (cbethell) [reply 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd:389]: I attempted to do this, but not sure I _fully_ followed the citation guidelines? Wasn't completely sure who the author should be (`CCDL 2020` as in the notebook or those who I can see contributed in Git history).

**Comment by cbethell** [comment]
I believe I addressed all of your comments @cansavvy (I went through it more times that I can tell you at this point so I surely hope I addressed all of your formatting comments). Please tell me what you think about the added citation however, which references our `training-modules` notebook:

```
I attempted to cite in the notebook, 
but not sure I fully followed the citation guidelines? 
Wasn't completely sure who the author should be 
(CCDL 2020 as in the notebook or those who I can see contributed in Git history).
```

File: 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd (hunk 12 above)
Comment (cansavvy) [reply 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd:389]: This is a tricky instance. I think what you have now is fine.

File: 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd [hunk 19]
```diff
@@ -0,0 +1,421 @@
...
+This means that you should not rely too heavily on the exact values of UMAP's output. 
+
+  - One particular limitation of UMAP is that while observed clusters have some meaning, the distance *between* clusters usually does not (nor does cluster density). The fact that two clusters are near each other should NOT be interpreted to mean that they are more related to each other than to more distant clusters. (There is some disagreement about whether UMAP distances have more meaning, but it is also probably safer to assume they don't.)
+
+2. Playing with the parameters so you can fine-tune them is a good way to give you more information about a particular analysis as well as the data itself.
+Feel free to try playing with the parameters on your own in the code chunks above!
+   
+In summary, a good rule of thumb to remember is: if the results of an analysis can be completely changed by changing its parameters, you should be very cautious when it comes to the conclusions you draw from it as well as having good rationale for the parameters you choose [@CCDL2020].
```
Comment (cansavvy) [inline 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd:394]:
```suggestion
In summary, a good rule of thumb to remember is: if the results of an analysis can be completely changed by changing its parameters, you should be very cautious when it comes to the conclusions you draw from it as well as having good rationale for the parameters you choose (_adapted from @CCDL2020 training materials_).
```

File: 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd [hunk 20]
```diff
@@ -0,0 +1,421 @@
...
+)
+```
+
+# Resources for further learning
+
+- [More on UMAP and its parameters](https://cran.r-project.org/web/packages/umap/vignettes/umap.html) [@umap-vignette]
+- [Guidelines on choosing dimension reduction methods](https://journals.plos.org/ploscompbiol/article/file?id=10.1371/journal.pcbi.1006907&type=printable) [@Nguyen2019]  
+- [A nice explanation and comparison of many different dimensionality reduction techniques that you may encounter](https://rpubs.com/Saskia/520216) [@Freytag2019]
+
+# Print session info:
```
Comment (cansavvy) [inline 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd:413]:
```suggestion
# Print session info
```

File: 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd [hunk 21]
```diff
@@ -0,0 +1,421 @@
...
+author: "CCDL for ALSF"
+date: "`r format(Sys.time(), '%B %Y')`"
+output:   
+  html_notebook:
+    toc: true
+    toc_float: true
+    number_sections: true
+---
+
+# Purpose of this analysis:
```
Comment (cansavvy) [inline 03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd:12]:
```suggestion
# Purpose of this analysis
```

File: references.bib [hunk 22]
```diff
@@ -1,3 +1,10 @@
+@Manual{CCDL2020,
+  title = {scRNA-seq Dimension Reduction},
+  author = {CCDL Team},
```
Comment (cansavvy) [inline references.bib:3]: One more comment
```suggestion
  author = {CCDL},
```
It thinks CCDL is the first name, Team is last name, so just delete the `Team` part.

**Pull request merged**